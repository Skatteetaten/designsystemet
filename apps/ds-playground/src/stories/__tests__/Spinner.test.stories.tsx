import { JSX } from 'react';

import { Meta, StoryObj } from '@storybook/react';
import { expect, within } from '@storybook/test';

import {
  Spinner,
  spinnerColorArr,
  spinnerPositionArr,
  spinnerSizeArr,
} from '@skatteetaten/ds-progress';

const meta = {
  component: Spinner,
  title: 'Tester/Spinner',
  argTypes: {
    // Baseprops
    key: { table: { disable: true } },
    ref: { table: { disable: true } },
    className: { table: { disable: true } },
    id: { table: { disable: true } },
    lang: { table: { disable: true } },
    'data-testid': { table: { disable: true } },
    // Props
    children: {
      table: { disable: true },
      control: 'text',
    },
    size: {
      table: { disable: true },
      control: 'inline-radio',
      options: [...spinnerSizeArr],
    },
    titlePosition: {
      table: { disable: true },
      control: 'inline-radio',
      options: [...spinnerPositionArr],
    },
    color: {
      table: { disable: true },
      control: 'inline-radio',
      options: [...spinnerColorArr],
    },
    classNames: { table: { disable: true } },
    percentComplete: { table: { disable: true } },
    hideTitle: {
      control: 'boolean',
      table: { disable: true },
    },
  },
} satisfies Meta<typeof Spinner>;
export default meta;
type Story = StoryObj<typeof meta>;

export const WithRef = {
  name: 'With Ref (FA1)',
  args: {
    ref: (instance: HTMLDivElement | null): void => {
      if (instance) {
        instance.id = 'dummyIdForwardedFromRef';
      }
    },
  },
  argTypes: {
    ref: { table: { disable: false } },
  },
  parameters: {
    imageSnapshot: { disable: true },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const spinner = canvas.getAllByRole('generic')[1];
    await expect(spinner).toHaveAttribute('id', 'dummyIdForwardedFromRef');
  },
} satisfies Story;

export const WithAttributes = {
  name: 'With Attributes (FA2-5)',
  args: {
    id: 'htmlid',
    className: 'dummyClassname',
    lang: 'nb',
    'data-testid': '123ID',
  },
  argTypes: {
    id: { table: { disable: false } },
    className: { table: { disable: false } },
    lang: { table: { disable: false } },
    'data-testid': { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const spinner = canvas.getByTestId('123ID');
    await expect(spinner).toHaveAttribute('id', 'htmlid');
    await expect(spinner).toHaveClass('dummyClassname');
    await expect(spinner).toHaveAttribute('lang', 'nb');
  },
} satisfies Story;

export const Defaults = {
  name: 'Defaults (B2)',
  argTypes: {
    children: { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const spinner = await canvas.findByText('Laster inn');
    await expect(spinner).toBeInTheDocument();
  },
} satisfies Story;

export const WithSizeAndPosition = {
  name: 'With Size, TitlePostition, Children (A1, A2, A4)',
  render: (args): JSX.Element => {
    return (
      <>
        <Spinner className={'bottomSpacingXL'} size={'small'} {...args} />
        <Spinner className={'bottomSpacingXL'} size={'medium'} {...args} />
        <Spinner className={'bottomSpacingXL'} size={'large'} {...args} />

        <Spinner
          className={'bottomSpacingXL'}
          titlePosition={'right'}
          size={'small'}
          {...args}
        />
        <Spinner
          className={'bottomSpacingXL'}
          titlePosition={'right'}
          size={'medium'}
          {...args}
        />
        <Spinner
          className={'bottomSpacingXL'}
          titlePosition={'right'}
          size={'large'}
          {...args}
        />
      </>
    );
  },
  args: {
    children: 'henter innhold',
  },
  argTypes: {
    children: { table: { disable: false } },
    size: { table: { disable: false } },
    titlePosition: { table: { disable: false } },
  },
} satisfies Story;

export const WithSizeAndPositionLongText = {
  ...WithSizeAndPosition,
  name: 'With Size, Text Postition, Children, Long Text (A1, A2, A4)',
  args: {
    children: 'Henter årsmiddelkurs fra valutakurstjenesten',
  },
};

export const WithColorsLightBackground = {
  name: 'With Color (A3)',
  render: (args): JSX.Element => {
    return (
      <>
        <Spinner
          className={'bottomSpacingXL'}
          color={'black'}
          size={'medium'}
          {...args}
        >
          {'Henter innhold'}
        </Spinner>
        <Spinner color={'blue'} size={'medium'} {...args}>
          {'Henter innhold'}
        </Spinner>
      </>
    );
  },
  argTypes: {
    color: {
      table: { disable: false },
    },
  },
} satisfies Story;

export const WithColorsDarkBackground = {
  name: 'With Color Dark Background(A3)',
  args: {
    color: 'white',
  },
  argTypes: {
    color: {
      table: { disable: false },
    },
  },
  parameters: {
    backgrounds: {
      default: 'dark',
    },
  },
} satisfies Story;

const spinnerTitle = 'laster data';
export const WithHideText = {
  name: 'With HideTitle (B1)',
  args: {
    hideTitle: true,
    children: spinnerTitle,
  },
  argTypes: {
    hideTitle: { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    // vi må bruke findBy i stedet for getBY her siden det er en liten delay før teksten rendres
    const spinnerTitleElement = await canvas.findByText(spinnerTitle);
    await expect(spinnerTitleElement).toBeInTheDocument();
  },
} satisfies Story;

export const WithPercent = {
  name: 'With PercentComplete (B3)',
  args: {
    percentComplete: 50,
    children: spinnerTitle,
  },
  argTypes: {
    percentComplete: { table: { disable: false } },
  },
  play: async ({ canvasElement }): Promise<void> => {
    const canvas = within(canvasElement);
    const spinner = canvas.getByRole('progressbar');
    await expect(canvas.getByText('50 %')).toBeInTheDocument();
    await expect(spinner).toHaveAttribute('aria-busy', 'true');
    await expect(spinner).toHaveAttribute('aria-valuenow', '50');
    await expect(spinner).toHaveAttribute('aria-valuemin', '0');
    await expect(spinner).toHaveAttribute('aria-valuemax', '100');
    await expect(spinner).toBeInTheDocument();
  },
} satisfies Story;
