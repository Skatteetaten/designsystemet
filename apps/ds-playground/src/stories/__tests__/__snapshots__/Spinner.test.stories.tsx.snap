// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tester/Spinner Defaults play-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25  "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Laster inn
    </span>
  </div>
</div>
`;

exports[`Tester/Spinner WithAttributes play-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       id="htmlid"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25 dummyClassname "
       lang="nb"
       data-testid="123ID"
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Laster inn
    </span>
  </div>
</div>
`;

exports[`Tester/Spinner WithColorsDarkBackground smoke-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_white_1imbs_19 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25  "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Laster inn
    </span>
  </div>
</div>
`;

exports[`Tester/Spinner WithColorsLightBackground smoke-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Henter innhold
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_blue_1imbs_16 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25  "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Henter innhold
    </span>
  </div>
</div>
`;

exports[`Tester/Spinner WithHideText play-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25  "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59 _srOnly_1imbs_70 _spinnerTitle_centerText_1imbs_66 ">
      laster data
    </span>
  </div>
</div>
`;

exports[`Tester/Spinner WithRef play-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25  "
       id="dummyIdForwardedFromRef"
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Laster inn
    </span>
  </div>
</div>
`;

exports[`Tester/Spinner WithSizeAndPosition smoke-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_small_1imbs_22 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      henter innhold
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      henter innhold
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_large_1imbs_28 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      henter innhold
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexRow_1imbs_36 _spinner_small_1imbs_22 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59   ">
      henter innhold
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexRow_1imbs_36 _spinner_medium_1imbs_25 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59   ">
      henter innhold
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexRow_1imbs_36 _spinner_large_1imbs_28 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59   ">
      henter innhold
    </span>
  </div>
</div>
`;

exports[`Tester/Spinner WithSizeAndPositionLongText smoke-test 1`] = `
<div data-test-block="true">
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_small_1imbs_22 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Henter årsmiddelkurs fra valutakurstjenesten
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_medium_1imbs_25 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Henter årsmiddelkurs fra valutakurstjenesten
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexColumn_1imbs_32 _spinner_large_1imbs_28 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59  _spinnerTitle_centerText_1imbs_66 ">
      Henter årsmiddelkurs fra valutakurstjenesten
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexRow_1imbs_36 _spinner_small_1imbs_22 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59   ">
      Henter årsmiddelkurs fra valutakurstjenesten
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexRow_1imbs_36 _spinner_medium_1imbs_25 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59   ">
      Henter årsmiddelkurs fra valutakurstjenesten
    </span>
  </div>
  <div aria-live="polite"
       aria-atomic="true"
       class="_spinner_1imbs_5 _spinner_black_1imbs_13 _spinner_flexRow_1imbs_36 _spinner_large_1imbs_28 bottomSpacingXL "
  >
    <div class="_spinnerAnimation_1imbs_40 ">
    </div>
    <span class="_spinnerTitle_1imbs_59   ">
      Henter årsmiddelkurs fra valutakurstjenesten
    </span>
  </div>
</div>
`;
